Code Navigation IA revamp (#202)
sourcegraph/docs: docs/code-search/code-navigation/auto_indexing.mdx

# Auto-indexing

<p className="subtitle">Learn and understand how auto-indexing works.</p>

<Callout type="note">This feature is in beta for self-hosted customers.</Callout>

> 

Reviewer: @mmanela Technically, auto-indexing is enabled by default for Cloud customers (c.f. related discussion on how many executors to make available to customers). However, we're missing these bullets from our [GA definition](https://handbook.sourcegraph.com/departments/engineering/sdlc/).
- Performance optimized.
- Come with scenario-complete docs + samples.
- Sales and support staff trained and ready to go.

Should we change this to 'in beta for both self-hosted and Cloud customers' or just 'in beta'?

With Sourcegraph deployments supporting [executors](/admin/executors/), your repository contents can be automatically analyzed to produce a code graph index file. Once [auto-indexing is enabled](/code_navigation/how-to/enable_auto_indexing) and [auto-indexing policies are configured](/code_navigation/how-to/configure_auto_indexing), repositories will be periodically cloned into an executor sandbox, analyzed, and the resulting index file will be uploaded back to the Sourcegraph instance.

## Enable Code Navigation auto-indexing

> <Callout type="note">This feature is in beta for self-hosted customers.</Callout>

Author: Nice catch.

The following docs explains how to turn on [auto-indexing](/code_navigation/explanations/auto_indexing) on your Sourcegraph instance to enable [precise code navigation](/code_navigation/explanations/precise_code_navigation).

### Deploy executors

> First, [deploy the executor service](/admin/executors/deploy_executors) targeting your Sourcegraph instance. This will provide the necessary compute resources that clone the target Git repository, securely analyze the code to produce a code graph data index, then upload that index to your Sourcegraph instance for processing.

Reviewer: @MaedahBatool how do we handle different steps for Cloud vs self-hosted customers in other parts of the docs? Here, the step of 'Deploy executors' only makes sense for self-hosted customers, we handle this ourselves for Cloud customers.

Author: @varungandhi-src, there is no set pattern for segregating docs for Cloud vs. Self-hosted audience. The best we can do is add a callout stating this is for self-hosted customers.

### Enable index job scheduling

> Next, enable the precise code navigation auto-indexing feature by enabling the following feature flag in your Sourcegraph instance's site configuration.

Reviewer: ```suggestion
Next, enable the following feature flag in your Sourcegraph instance's site configuration.
```

### Configure auto-indexing policies

Once auto-indexing has been enabled, [create auto-indexing policies](/code_navigation/how-to/configure_auto_indexing) to control the set of repositories and commits that are eligible for indexing. Note that policies only select repositories and branches/tags/commits that can be indexed, not how they are indexed.

### Configure auto-indexing jobs

For repository and commit pairs that are marked as eligible for indexing, the index job is either [inferred from the project](/code_navigation/explanations/auto_indexing_inference), or [explicitly configured](/code_navigation/how-to/configure_auto_indexing#explicit-index-job-configuration) on a per-repository basis.

### Tune the index scheduler

The frequency of index job scheduling can be tuned via the following environment variables read by `worker` service containers running the [`codeintel-auto-indexing`](/admin/workers#codeintel-auto-indexing) task.

- **`PRECISE_CODE_INTEL_AUTO_INDEXING_TASK_INTERVAL`**: The frequency with which to run periodic codeintel auto-indexing tasks. Default is every 2 minutes

> - **`PRECISE_CODE_INTEL_AUTO_INDEXING_REPOSITORY_PROCESS_DELAY`**: The minimum frequency that the same repository can be considered for auto-index scheduling. Default is every 24 hours

Reviewer: The wording here is confusing because the variables correspond to time durations, and the description says "frequency" which is technically the reciprocal of the time duration between consecutive iterations.

- **`PRECISE_CODE_INTEL_AUTO_INDEXING_REPOSITORY_BATCH_SIZE`**: The number of repositories to consider for auto-indexing scheduling at a time. Default is 2500
- **`PRECISE_CODE_INTEL_AUTO_INDEX_MAXIMUM_REPOSITORIES_INSPECTED_PER_SECOND`**: The maximum number of repositories inspected for auto-indexing per second. Set to zero to disable limit. Default is 0

### Access to private repositories and packages

> Auto-indexing jobs run as Docker containers on the Executors infrastructure, and by default don't have any knowledge about the private repositories or package registries your builds might use.

Reviewer: ```suggestion
Auto-indexing jobs run as Docker containers on the executors infrastructure, and by default don't have any knowledge about the private repositories or package registries your builds might use.
```

Earlier in the doc, "executors" is spelled with a lowercase "e".

The type of information you will need to provide depends on the language and ecosystem, but the mechanism is the same - [adding a secret to Executors "Code Graph" namespace](/admin/executors/executor_secrets).

See the [language-specific instructions](/code_navigation/how-to/ configure_auto_indexing#private-repositories-and-packages-configuration)

## Configure Code Navigation auto-indexing

<Callout type="info">This feature is in beta for self-hosted customers.</Callout>

> Precise code navigation [auto-indexing](/code_navigation/explanations/auto_indexing) jobs are scheduled based on two fronts of configuration.

Reviewer: The inconsistent use of 'auto-indexing' vs 'Code navigation auto-indexing' vs 'Precise code navigation auto-indexing' is quite distracting. Also, we will soon be adding a new form of auto-indexing, which will not be precise, so we have been calling it 'syntactic auto-indexing' internally.

For that, _most_ of the settings will be shared with precise auto-indexing, with some minor tweaks. So I propose simplifying the terminology to only use 'auto-indexing' for now. For the future, in places we need a finer distinction, we can use 'precise auto-indexing' and 'syntactic auto-indexing' as shorthand for 'auto-indexing for precise code navigation' and 'auto-indexing for syntactic code navigation respectively'.

Author: Nice callout and I agree here. Let's simplify it, and maybe we can change it later.

The first front selects the set of repositories and commits within those repositories that are candidates for auto-indexing. These candidates are controlled by [configuring auto-indexing policies](#configure-auto-indexing-policies).

The second front determines the set of index jobs that can run over candidate commits. By default, index jobs are [inferred](/code_navigation/explanations/auto_indexing_inference) from the repository structure's on disk. Index job inference uses heuristics such as the presence or contents of particular files to determine the paths and commands required to index a repository.

> Alternatively, index job configuration [can be supplied explicitly](#explicit-index-job-configuration) for a repository when the inference heuristics are not powerful enough to create an index job that produces the correct results. This might be necessary for projects that have non-standard or complex dependency resolution or pre-compilation steps, for example.

Reviewer: It seems like there is a bunch of overlap between this section and the sub-sections earlier:

```
### Configure auto-indexing policies

...

### Configure auto-indexing jobs
```

What's the reason for that? Maybe the earlier sub-sections should be removed?

Author: Makes sense. Let's remove redundant content.

### Configure auto-indexing policies

> Let's learn how to configure policies to control the scheduling of precise code navigation indexing jobs. Indexing jobs [produce a code graph data index](/code_navigation/explanations/precise_code_navigation) and uploads it to your Sourcegraph instance for use with code navigation.

Reviewer: ```suggestion
Let's learn how to configure policies to control the scheduling of precise code navigation indexing jobs. An indexing jobs [produces a code graph data index](/code_navigation/explanations/precise_code_navigation) and uploads it to your Sourcegraph instance for use with code navigation.
```

Since the verb in the second clause is 'uploads', having 'indexing jobs' earlier is incongruous.

<Callout type="note">See the [best practices guide](/code_navigation/how-to/policies_resource_usage_best_practices) for additional details on how policies affect resource usage.</Callout>

Each policy has a number of configurable options, including:

- The set of Git branches or tags to which the policy applies
- The maximum age of commits that should be indexed (e.g., skip indexing commits made last year)

> - For branches, whether or not to consider the _tip_ of the branch only, or all commits contained in that branch

Reviewer: ```suggestion
- For branches, whether to only consider the _tip_ of the branch, or all commits on that branch
```

> Note that when auto-indexing is enabled, we will also attempt to schedule index jobs for dependencies of repositories which receive an uploaded code graph data index. This helps to ensure that no matter where symbols are defined, you will be able to navigate to its definition and find a relevant set of references as long as your Sourcegraph instance has knowledge of that code.

Reviewer: ```suggestion
When auto-indexing is enabled, uploading a code graph data index will trigger creation of index jobs for dependencies if they are available on the Sourcegraph instance, in order to provide greater coverage for precise code navigation actions such as Go to definition and Find references.
```
